monitoring/exporter/exporter.go

```diff
@@ -0,0 +1,232 @@
+
+	// mu protects access to projDataMap
+	mu sync.Mutex
+	// per-project data of exporter
+	projDataMap map[string]*projectData
+}
+
+// Options designates various parameters used by stats exporter. Default value of fields in Options
+// are valid for use.
+type Options struct {
+	// ClientOptions designates options for creating metric client, especially credentials for
+	// RPC calls.
+	ClientOptions []option.ClientOption
+
+	// options for bundles amortizing export requests. Note that a bundle is created for each
+	// project. When not provided, default values in bundle package are used.
+	BundleDelayThreshold time.Duration
+	BundleCountThreshold int
+
+	// callback functions provided by user.
+
+	// GetProjectID is used to filter whether given row data can be applicable to this exporter
+	// and if so, it also determines the projectID of given row data. If
+	// RowDataNotApplicableError is returned, then the row data is not applicable to this
+	// exporter, and it will be silently ignored. Though not recommended, other errors can be
+	// returned, and in that case the error is reported to callers via OnError and the row data
+	// will not be uploaded to stackdriver. When GetProjectID is not set, all row data will be
+	// considered not applicable to this exporter.
+	GetProjectID func(*RowData) (projectID string, err error)
+	// OnError is used to report any error happened while exporting view data fails. Whenever
+	// this function is called, it's guaranteed that at least one row data is also passed to
+	// OnError. Row data passed to OnError must not be modified. When OnError is not set, all
+	// errors happened on exporting are ignored.
+	OnError func(error, ...*RowData)
+	// MakeResource creates monitored resource from RowData. It is guaranteed that only RowData
+	// that passes GetProjectID will be given to this function. Though not recommended, error
+	// can be returned, and in that case the error is reported to callers via OnError and the
+	// row data will not be uploaded to stackdriver. When MakeResource is not set, global
+	// resource is used for all RowData objects.
+	MakeResource func(rd *RowData) (*monitoredrespb.MonitoredResource, error)
+
+	// options concerning labels.
+
+	// DefaultLabels store default value of some labels. Labels in DefaultLabels need not be
+	// specified in tags of view data. Default labels and tags of view may have overlapping
+	// label keys. In this case, values in tag are used. Default labels are used for labels
+	// those are constant throughout export operation, like version number of the calling
+	// program.
+	DefaultLabels map[string]string
+	// UnexportedLabels contains key of labels that will not be exported stackdriver. Typical
+	// uses of unexported labels will be either that marks project ID, or that's used only for
+	// constructing resource.
+	UnexportedLabels []string
+}
+
+// default values for options
+func defaultGetProjectID(rd *RowData) (string, error) {
```

Comment:
Would it better if the default function handled the checking of Tags in the RowData to see if there is a project_id tag and returning the corresponding ID?

Reply:
ok. I did and wrote a test for it